google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/MergedResultSet.java

```diff
@@ -0,0 +1,399 @@
+
+/**
+ * {@link MergedResultSet} is a {@link ResultSet} implementation that combines the results from
+ * multiple queries. Each query uses its own {@link RowProducer} that feeds rows into the {@link
+ * MergedResultSet}. The order of the records in the {@link MergedResultSet} is not guaranteed.
+ */
+class MergedResultSet extends ForwardingStructReader implements PartitionedQueryResultSet {
+  static class PartitionExecutor implements Runnable {
+    private final Connection connection;
+    private final String partitionId;
+    private final LinkedBlockingDeque<PartitionExecutorResult> queue;
+    private final AtomicBoolean shouldStop = new AtomicBoolean();
+
+    PartitionExecutor(
+        Connection connection,
+        String partitionId,
+        LinkedBlockingDeque<PartitionExecutorResult> queue) {
+      this.connection = Preconditions.checkNotNull(connection);
+      this.partitionId = Preconditions.checkNotNull(partitionId);
+      this.queue = queue;
+    }
+
+    @Override
+    public void run() {
+      try (ResultSet resultSet = connection.runPartition(partitionId)) {
+        boolean first = true;
+        while (resultSet.next()) {
+          Struct row = resultSet.getCurrentRowAsStruct();
+          if (first) {
+            queue.put(
+                PartitionExecutorResult.dataAndMetadata(
+                    row, resultSet.getType(), resultSet.getMetadata()));
+            first = false;
+          } else {
+            queue.put(PartitionExecutorResult.data(row));
+          }
+          if (shouldStop.get()) {
+            break;
+          }
+        }
+        if (first) {
+          // Special case: The result set did not return any rows. Push the metadata to the merged
+          // result set.
+          queue.put(
+              PartitionExecutorResult.typeAndMetadata(
+                  resultSet.getType(), resultSet.getMetadata()));
+        }
+      } catch (Throwable exception) {
+        putWithoutInterruptPropagation(PartitionExecutorResult.exception(exception));
+      } finally {
+        // Emit a special 'finished' result to ensure that the row producer is not blocked on a
+        // queue that never receives any more results. This ensures that we can safely block on
+        // queue.take(), as we know that we will always receive at least one result from each
+        // worker.
+        putWithoutInterruptPropagation(PartitionExecutorResult.finished());
+      }
+    }
+
+    private void putWithoutInterruptPropagation(PartitionExecutorResult result) {
+      try {
+        queue.put(result);
+      } catch (InterruptedException interruptedException) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  static class PartitionExecutorResult {
+    private final Struct data;
+    private final Throwable exception;
+    private final Type type;
+    private final ResultSetMetadata metadata;
+
+    static PartitionExecutorResult data(Struct data) {
```

Comment:
Wouldn't a builder pattern be better suited for such object construction? Otherwise for every member that we introduce it will be difficult to multiplex such method to partially build the object.